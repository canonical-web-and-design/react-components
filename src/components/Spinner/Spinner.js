import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames";

import "./Spinner.scss";

const Spinner = ({ className, text, isLight, inline }) => (
  <span
    className={classNames(className, "p-spinner", "p-text--default", {
      "p-spinner--inline": inline
    })}
  >
    <i
      className={classNames("p-icon--spinner", "u-animation--spin", {
        "is-light": isLight
      })}
    />
    {text && <span className="p-icon__text">{text}</span>}
  </span>
);

Spinner.propTypes = {
  className: PropTypes.string,
  text: PropTypes.string,
  isLight: PropTypes.bool,
  inline: PropTypes.bool
};

export default Spinner;
