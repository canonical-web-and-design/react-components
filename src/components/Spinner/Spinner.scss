@import "~vanilla-framework/scss/settings_spacing";

.p-icon__text {
  margin-left: $sph-inner--small;
}

.p-spinner {
  line-height: map-get($line-heights, default-text);
  padding: ($spv-nudge - 0.0625rem) ($sph-inner--small * 1.5);
  text-align: center;
  width: 100%;

  button & {
    padding: 0;
  }
}

.p-spinner--inline {
  display: inline;
  margin: 0 1rem;
  text-align: left;
}
