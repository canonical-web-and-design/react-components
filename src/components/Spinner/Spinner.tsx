import React from "react";
import classNames from "classnames";

export type Props = {
  /**
   * Optional class(es) to pass to the wrapping span element.
   */
  className?: string;
  /**
   * Whether the spinner should have a light appearance.
   */
  isLight?: boolean;
  /**
   * Text to display next to the spinner.
   */
  text?: string;
};

const Spinner = ({ className, text, isLight = false }: Props): JSX.Element => (
  <span className={classNames(className, "p-text--default")}>
    <i
      className={classNames("p-icon--spinner", "u-animation--spin", {
        "is-light": isLight,
      })}
    />
    {text && (
      <>
        &ensp;
        <span>{text}</span>
      </>
    )}
  </span>
);

export default Spinner;
