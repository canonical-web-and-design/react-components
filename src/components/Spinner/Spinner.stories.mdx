import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";
import { withKnobs } from "@storybook/addon-knobs";
import Spinner from "./Spinner";

<Meta title="Spinner" component={Spinner} decorators={[withKnobs]} />

### Spinner

This is a [React](https://reactjs.org/) component for the Vanilla [Spin](https://docs.vanillaframework.io/settings/animation-settings/#spin) animation.

### Props

<ArgsTable of={Spinner} />

### Default

<Canvas>
  <Story name="Default">
    <Spinner text="Loading..." />
  </Story>
</Canvas>
