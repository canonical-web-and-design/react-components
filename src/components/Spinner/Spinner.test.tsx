import { shallow } from "enzyme";
import React from "react";

import Spinner from "./Spinner";

describe("<Spinner />", () => {
  // snapshot tests
  it("renders and matches the snapshot", () => {
    const component = shallow(<Spinner />);

    expect(component).toMatchSnapshot();
  });

  it("renders and matches the snapshot with text", () => {
    const component = shallow(<Spinner text="foo" />);

    expect(component).toMatchSnapshot();
  });

  // unit tests
  it("renders a spinner icon", () => {
    const component = shallow(<Spinner />);

    expect(
      component.find("i").first().hasClass("p-icon--spinner")
    ).toBeTruthy();
  });

  it("renders text correctly if given text prop", () => {
    const text = "Loading...";
    const component = shallow(<Spinner text={text} />);

    expect(component.find("span").first().text()).toEqual(text);
  });

  it("renders a light spinner if given isLight prop", () => {
    const component = shallow(<Spinner isLight />);

    expect(component.find("i").first().hasClass("is-light")).toEqual(true);
  });
});
