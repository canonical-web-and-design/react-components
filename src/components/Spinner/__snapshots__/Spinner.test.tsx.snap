// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> renders and matches the snapshot 1`] = `
<span
  className="p-spinner p-text--default"
>
  <i
    className="p-icon--spinner u-animation--spin"
  />
</span>
`;

exports[`<Spinner /> renders and matches the snapshot with text 1`] = `
<span
  className="p-spinner p-text--default"
>
  <i
    className="p-icon--spinner u-animation--spin"
  />
  <span
    className="p-icon__text"
  >
    foo
  </span>
</span>
`;
